src/execution/execute.js

```diff
@@ -977,6 +1022,47 @@ function completeAbstractValue(
 }
 
 /**
+ * If a resolveType function is not given, then a default resolve behavior is
+ * used which tests each possible type for the abstract type by calling
+ * isTypeOf for the object being coerced, returning the first type that matches.
+ */
+function resolveFirstValidType(
+  value: mixed,
+  context: mixed,
+  info: GraphQLResolveInfo,
+  possibleTypes: Array<GraphQLObjectType>,
+  i: number = 0,
```

Comment:
Done. Here are some thoughts on the new implementation.
- The ordering of types is not preserved when mixing sync and async `isTypeOf`:
```js
const Dog = new GraphQLObjectType({
  isTypeOf: () => Promise.resolve(true),
  /* ... */
});

const Cat = new GraphQLObjectType({
  isTypeOf: () => true,
  /* ... */
});

const PetType = new GraphQLUnionType({
  types: [Dog, Cat]
  /* ... */
});
```
In a fully-sync schema, Dog would be picked as the correct type for any query on Pet, but in the proposed implementation, we resolve the sync ones first, so Cat is picked. In a fully async schema the ordering is preserved. (Keeping that ordering was what I was trying to achieve using the outdated recursive function)
- Error handling changed a bit. See [this diff](https://github.com/graphql/graphql-js/pull/631/commits/ae5295a2d3b3b19aa3e45129ac1a18b163ca9aea#diff-ad7767caa77907dedfe8013eaad8ff86):
```js
const DogType = new GraphQLObjectType({
  isTypeOf: source => Promise.resolve(source instanceof Dog),
  /* ... */
});

const CatType = new GraphQLObjectType({
  isTypeOf: () => Promise.reject('Oh no!'),
  /* ... */
});

const PetType = new GraphQLUnionType({
  types: [Dog, Cat]
  /* ... */
});
```
When querying for a Dog everything should be alright but since we run `Promise.all` on all async `isTypeOf` functions it throws the Cat error. Maybe that is not an issue since rejecting in `isTypeOf` would never be the developer's intent.